_posts/2021-01-04-Designing-REST-services.md

```diff
@@ -0,0 +1,331 @@
+
+There are several considerations that way on the design and implementation of services that are of a more technical nature. 
+Some of them have already been mentioned, but there is a selection of these characteristics that are worked out in the book, each with its proper impact on the scalability of the services.
+
+### Caching
+
+Caching can be done at numerous points along the request/response chain of a service call. 
+When a resource is requested, and one of the intermediate components has a version of this resource, they will provide it back. 
+Otherwise the service call will reach the origin service and collect the real-time data. 
+The origin service should provide the intermediate components the rules for when and how to cache as well as how long the data would be considered “fresh”. 
+Evidently, the closer the caching component is to the consumer in the request/response chain, the less expensive (reduction of bandwidth and of load on the origin service) the call. 
+The other benefits of caching are a reduced latency (a quicker roundtrip time) and a reduced impact of possible network failures. 
+
+While there are benefits to caching, there are also reasons not to do so. 
+In these situations, caching will harm more than it benefits, or might simply not be allowed:
+* If the GET of the resource generates side effects from being accessed (for example a counter system that limits how many times the resource may be requested by consumers or some early bird system that benefits the first so many consumers), the cache would prevent these side effects.
+* When the system in place doesn’t tolerate any latency in data to be retrieved and we need to be sure that the data received is real-time (for example a heart monitor in ICU wards).
+* When caching is not allowed for regulatory reasons or security/privacy risks.
+* When the frequency with which the data changes is so high and the period that the data is not stale is so small that caching would never trigger.
+
+Aside from all the caching that can be added on an applicative level, there are already several components capable of caching that are native to the infrastructure of the web:
+* Local Cache: A store of cached representations from many origin servers at the behest of a single consumer, either in memory or persisted.
+* Proxy Cache: A server that stores representations from many origin servers to many different consumers. This component can reside either outside or inside the corporate firewall.
+* Reverse Proxy: A component that is a type of proxy server that retrieves resources from a single origin server on behalf of any number of consumers. It focuses mainly on load balancing, HTTP acceleration and security features. 
+
+The caching components are instructed on how to handle the caching of the resources they hold by using these specific HTTP response headers:
+* Expires: Indicates the length of the period the cached resource can be used before being considered stale.
```

Comment:
Maybe use backticks on all headers for better readability.

Reply:
done